README.md
# SensuPluginsZFS

Sensu plugin for zfs health checks.

[![Build Status](https://travis-ci.org/sensu-plugins/sensu-plugins-zfs.svg?branch=master)](https://travis-ci.org/sensu-plugins/sensu-plugins-zfs)
[![Gem Version](https://badge.fury.io/rb/sensu-plugins-zfs.svg)](http://badge.fury.io/rb/sensu-plugins-zfs)
[![Dependency Status](https://gemnasium.com/sensu-plugins/sensu-plugins-zfs.svg)](https://gemnasium.com/sensu-plugins/sensu-plugins-zfs)
[![Community Slack](https://slack.sensu.io/badge.svg)](https://slack.sensu.io/badge)

## Checks

### check-zpool.rb

This checks does the following checks against one or more zpools.

- Check state
- Check vdev errors
- Check capacity
- Check that a scrub has been run recently

The following flags can be used to configure the checks.

- `-z, --zpool` What zpool to check. If omitted, we check all zpools.
- `-c, --capacity-warn` Capacity threshold for when to warn. (default 80)
- `-C, --capacity-crit` Capacity threshold for when to crit. (default 90)
- `-s, --scrubbing-interval` Warn when it is more than this number of days since last scrub. (default 7)

## Metrics

### metric-zfs-arc.rb

Reads `/proc/spl/kstat/zfs/arcstats` for ARC statistics and puts them in a form usable by Graphite.

For more information see ZoL [Linux Module Parameters](https://github.com/zfsonlinux/zfs/wiki/ZFS-on-Linux-Module-Parameters)

## Installation
[Installation and setup](http://sensu-plugins.io/docs/installation_instructions.html)

## Contributing

At this time ideas for additional checks/metrics would be very much appreciated.

I have a few ideas that would be nice:

- [x] Check for zpool state
- [x] Check for vdev errors
- [x] Check for pool capacity
- [x] Check that disks have been scrubbed recently
- [ ] Metric for disk utilization

Bug reports and pull requests are welcome on GitHub at https://github.com/sensu-plugins/sensu-plugins-zfs.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows the format located [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)


## [Unreleased] - 2020-02-10

## [2.2.0] - 2020-03-02
### Changed
- release adds support for sensu-go/bonsai
- Update rubocop to '~> 0.80.1'
- Remediate rubocop items
- Updated rake requirement from 12.3 to ~> 13.0

## [2.1.0] - 2019-05-06
### Added
- metric-zfs-arc.rb, aggregate ZFS ARC metrics (@phumpal)

## [2.0.1] - 2018-02-27
### Changed
- this is a no-op release to test the CI pipeline's ability to publish gems

## [2.0.0] - 2018-02-27
### Security
- updated rubocop dependency to `~> 0.51.0` per: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-8418 (@majormoses)

### Breaking Changes
- in order to bring in the newer rubocop we had to drop ruby `< 2.1` support (@majormoses)

### Changed
- accepted into community, a bunch of initial house keeping (@majormoses)

### Fixed
- fix scrub-in-progress command
- properly parse the list of vdevs
- fix checking all zpools (#2)

[Unreleased]: https://github.com/sensu-plugins/sensu-plugins-zfs/compare/2.2.0...HEAD
[2.2.0]: https://github.com/sensu-plugins/sensu-plugins-zfs/compare/2.1.0...2.2.0
[2.1.0]: https://github.com/sensu-plugins/sensu-plugins-zfs/compare/2.0.1...2.1.0
[2.0.1]: https://github.com/sensu-plugins/sensu-plugins-zfs/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/sensu-plugins/sensu-plugins-zfs/compare/bf20f6b2538849a9263dbaa8771d649b7173d8b1...2.0.0


spec/spec_helper.rb
require "bundler/setup"
require "sensu/plugins/zfs"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/sensu/plugins/zfs_spec.rb
require "spec_helper"

RSpec.describe Sensu::Plugins::Zfs do
  it "has a version number" do
    expect(Sensu::Plugins::Zfs::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


lib/sensu-plugins-zfs/zpool.rb
# frozen_string_literal: true

require 'time'

module SensuPluginsZFS
  class ZFS
    def self.zpools
      `sudo zpool list -H -o name`.lines.map do |l|
        ZPool.new(l.strip)
      end
    end
  end

  class ZPool
    attr_reader :name, :state, :capacity, :vdevs

    def initialize(name)
      @name = name
      @state = `sudo zpool status #{name} | grep '^ state: ' | cut -d ' ' -f 3`.strip
      @capacity = `sudo zpool get -H capacity #{@name} | awk '{print $3}' | cut -d '%' -f1`.strip.to_i
      @vdevs = create_vdevs name
    end

    def ok?
      @state == 'ONLINE'
    end

    def scrubbed_at
      return Time.at(0) if never_scrubbed?
      return Time.now if scrub_in_progress?

      Time.parse `sudo zpool status #{@name} | grep '^  scan: scrub' | awk '{print $11" "$12" "$13" "$14" "$15}'`.strip # rubocop:disable
    end

    private

    def create_vdevs(name)
      cmd_out = `sudo zpool status #{name} | grep ONLINE | grep -v state | awk '{print $1 " " $2 " " $3 " " $4 " " $5}'`
      cmd_out.lines.map do |l|
        arr = l.strip.split(' ')
        VDev.new(self, arr[0], arr[1], arr[2].to_i, arr[3].to_i, arr[4].to_i)
      end
    end

    def never_scrubbed?
      `sudo zpool status #{@name} | egrep -c "none requested"`.strip.to_i == 1
    end

    def scrub_in_progress?
      `sudo zpool status #{@name} | egrep -c "scrub in progress|resilver"`.strip.to_i == 1
    end
  end

  class VDev
    attr_reader :zpool, :name, :read, :write, :chksum
    def initialize(zpool, name, state, read, write, chksum) # rubocop:disable Metrics/ParameterLists
      @zpool = zpool
      @name = name
      @state = state
      @read = read
      @write = write
      @chksum = chksum
    end

    def ok?
      # TODO: I am not sure why this double negation exists and someone should
      # come back in and fix this later.
      !!(@state =~ /(ONLINE|AVAIL)/) && # rubocop:disable Style/DoubleNegation:
        @read.zero? &&
        @write.zero? &&
        @chksum.zero?
    end
  end
end


lib/sensu-plugins-zfs/zfs.rb
require 'time'

module SensuPluginsZFS
  class ZFS
    def self.zpools
      `sudo zpool list -H -o name`.lines.map do |l|
        ZPool.new(l.strip)
      end
    end

    def self.zfslist
      `sudo zfs list -H -o name`.lines.map do |l|
        ZFSList.new(l.strip)
      end
    end
  end

  class ZFSList
    attr_reader :name, :used, :available, :quota

    def initialize(name)
      @name = name
      @used = `sudo zfs get -Hp used "#{name}" | awk '{print $3}'`.strip.to_i
      @available = `sudo zfs get -Hp available "#{name}" | awk '{print $3}'`.strip.to_i
      @quota = `sudo zfs get -Hp quota "#{name}" | awk '{print $3}'`.strip.to_i
    end

    def calc_percentage_quota
      return ((@used * 100) / @quota) if quota?
      0
    end

    def quota?
      @quota > 0
    end
  end

  class ZPool
    attr_reader :name, :state, :capacity, :vdevs

    def initialize(name)
      @name = name
      @state = `sudo zpool status #{name} | grep '^ state: ' | cut -d ' ' -f 3`.strip
      @capacity = `sudo zpool get -H capacity #{@name} | awk '{print $3}' | cut -d '%' -f1`.strip.to_i
      @vdevs = create_vdevs name
    end

    def ok?
      @state == 'ONLINE'
    end

    def scrubbed_at
      return Time.at(0) if never_scrubbed?
      return Time.now if scrub_in_progress?
      Time.parse `sudo zpool status #{@name} | grep '^  scan: scrub' | awk '{print $11" "$12" "$13" "$14" "$15}'`.strip # rubocop:disable
    end

    private

    def create_vdevs(name)
      cmd_out = `sudo zpool status #{name} | grep ONLINE | grep -v state | awk '{print $1 " " $2 " " $3 " " $4 " " $5}'`
      cmd_out.lines.map do |l|
        arr = l.strip.split(' ')
        VDev.new(self, arr[0], arr[1], arr[2].to_i, arr[3].to_i, arr[4].to_i)
      end
    end

    def never_scrubbed?
      `sudo zpool status #{@name} | egrep -c "none requested"`.strip.to_i == 1
    end

    def scrub_in_progress?
      `sudo zpool status #{@name} | egrep -c "scrub in progress|resilver"`.strip.to_i == 1
    end
  end

  class VDev
    attr_reader :zpool, :name, :read, :write, :chksum
    def initialize(zpool, name, state, read, write, chksum) # rubocop:disable Metrics/ParameterLists
      @zpool = zpool
      @name = name
      @state = state
      @read = read
      @write = write
      @chksum = chksum
    end

    def ok?
      # TODO: I am not sure why this double negation exists and someone should
      # come back in and fix this later.
      !!(@state =~ /(ONLINE|AVAIL)/) && # rubocop:disable Style/DoubleNegation:
        @read.zero? &&
        @write.zero? &&
        @chksum.zero?
    end
  end
end
